### PR TITLE
Multichannel beat

### PR DESCRIPTION
**Pull request by iranroman**
#### Reference Issue

Implements #1130

#### What does this implement/fix? Explain your changes.
work in progress for multichannel beat

### CODE REVIEW

**Comment by iranroman** [comment]
Sharing the changes I've done towards the multichannel beat module. So far I've written the onset strength function and corresponding test. Looking forward to your comments and advice.

**Comment by bmcfee** [comment]
Can you open this pr against the multichannel branch instead? We haven't merged to main yet, so this pr is showing a bunch of unrelated changes.

**Comment by bmcfee** [comment]
Ah, it turns out that I could change the PR target branch from the web interface (but not the app), so disregard the previous comment.

That said, you have conflicts, so please resolve those first.

**Comment by iranroman** [comment]
> That said, you have conflicts, so please resolve those first.

Thanks for the heads up. It was just a missing `optimize=True` that I did not have in my temporary multichannel melspectrogram implementation.

File: librosa/onset.py [hunk 1]
```diff
@@ -345,7 +345,7 @@ def onset_strength(
         **kwargs,
     )
 
-    return odf_all[0]
+    return np.squeeze(odf_all)
```
Comment (bmcfee) [inline librosa/onset.py:348]: This seems dangerous to me.  `np.squeeze` removes all singleton dimensions, where the old code was only removing one dimension.

Would `return odf_all[..., 0, :]` work instead?

File: librosa/onset.py [hunk 2]
```diff
@@ -602,22 +602,27 @@ def onset_strength_multi(
...
     pad_width = lag
     if center:
         # Counter-act framing effects. Shift the onsets by n_fft / hop_length
         pad_width += n_fft // (2 * hop_length)
 
-    onset_env = np.pad(onset_env, ([0, 0], [int(pad_width), 0]), mode="constant")
+    onset_env_ndims = onset_env.ndim
+    onset_env = np.pad(onset_env, 
+                    [[0,0] if i != (onset_env_ndims-1) else [int(pad_width),0] 
+                        for i in range(onset_env_ndims)],
```
Comment (bmcfee) [inline librosa/onset.py:616]: I'd prefer to expand this list comprehension out into something more readable:
```python
padding = [(0,0) for _ in onset_env.shape]
padding[-1] = (int(pad_width), 0)
onset_env = np.pad(onset_env, padding, mode='constant')
```

We use this pattern elsewhere, and it's a bit more readable IMO.

File: tests/test_multichannel.py [hunk 3]
```diff
@@ -59,6 +92,25 @@ def test_stft_multi(y_multi):
...
+
+    D0 = librosa.onset.onset_strength(y[0])
+    D1 = librosa.onset.onset_strength(y[1])
+
+    # Check each channel
+    assert np.allclose(D[0], D0, atol=1e-1, rtol=1e-1) # test in dB units
+    assert np.allclose(D[1], D1, atol=1e-1, rtol=1e-1) # test in dB units
+
+    # Check that they're not both the same
+    assert not np.allclose(D0, D1, atol=1e-1, rtol=1e-1) # test in dB units
```
Comment (bmcfee) [inline tests/test_multichannel.py:111]: These tolerances are way loose.  How far off are the results actually?  (Also I don't think they're in dB)

Comment (iranroman) [reply tests/test_multichannel.py:111]: mmhmm, then there's something I do not understand here: 
https://github.com/librosa/librosa/blob/fad5538330312635a7ac91a9b9bb92d078b16593/librosa/onset.py#L571

Here are some general stats on the comparison being made during the test.

```
np.abs(D[0] - D0)
min: 0.0
max: 0.14240837
mean: 0.008507246

np.abs(D[1] - D1)
min: 0.0
max: 1.4305115e-06
mean: 7.587927e-08


np.abs(D0 - D1)
min: 0.0
max: 0.44862747
mean: 0.100642346
```

Comment (bmcfee) [reply tests/test_multichannel.py:111]: Ah, right.  This is another instance of the `power_to_db` calculation being sensitive to all channels simultaneously, as noted here: https://github.com/librosa/librosa/pull/1351#issuecomment-894711524

TLDR: it truncates to 80dB below peak, but the peak finder operates across all dimensions.  (I think this is correct behavior, and what we want most of the time.)

The fix here, I think, would be to use spectrogram input directly instead of implicitly computing a melspec by providing the time-domain input.

**Review by bmcfee — CHANGES_REQUESTED** [review]

File: librosa/core/spectrum.py [hunk 4]
```diff
@@ -245,6 +245,7 @@ def stft(
 
     if dtype is None:
         dtype = util.dtype_r2c(y.dtype)
+        print(dtype)
```
[the remark this answers is not in the document]
Comment (iranroman) [reply librosa/core/spectrum.py:248]: hehehe, oops (busted!)

File: librosa/beat.py [hunk 5]
```diff
@@ -374,7 +374,7 @@ def tempo(
 
     # Get the maximum, weighted by the prior
     # Using log1p here for numerical stability
-    best_period = np.argmax(np.log1p(1e6 * tg) + logprior[:, np.newaxis], axis=0)
+    best_period = np.argmax(np.log1p(1e6 * tg) + logprior[:, np.newaxis], axis=-2)
 
     return bpms[best_period]
```
[the remark this answers is not in the document]
Comment (iranroman) [reply librosa/beat.py:379]: Done, but not clear to me whether you mean `best_period` may be a multi-element vector (instead of just a list with one scalar), or something with more dimensions (i.e. matrix or ndarray).

[the remark this answers is not in the document]
Comment (bmcfee) [reply librosa/beat.py]: > or something with more dimensions (i.e. matrix or ndarray).

Yes: if you do `np.argmax(X, axis=n)`, the result will have shape equal to `X.shape` except dimension `n` is removed.  

In our case, we used to have `tg.ndim=2`, so argmaxing results in a 1d array, so `best_period` has shape `(n_frames,)`.  When we expand out to multichannel, we'll be collapsing out the tempo dimension, but retaining channels and frames.

My comment here comes from the concern that using a multidimensional index array can sometimes do unexpected things (see [here](https://numpy.org/doc/stable/reference/arrays.indexing.html#advanced-indexing)).  However, in this case we're okay because `bpms` is always 1d, so we don't get into any of the weird cases.

File: librosa/beat.py [hunk 6]
```diff
@@ -371,12 +371,14 @@ def tempo(
     if max_tempo is not None:
         max_idx = np.argmax(bpms < max_tempo)
         logprior[:max_idx] = -np.inf
+    # explicit axis expansion
+    logprior = np.expand_dims(logprior, axis=(-1, *range(tg.ndim-2)))
```
Comment (bmcfee) [inline librosa/beat.py:375]: Fancy :monocle_face:

But this relies on deprecated functionality in expand_dims when `tg.ndim-2 > logprior.ndim`.  See the deprecation note [here](https://numpy.org/doc/stable/reference/generated/numpy.expand_dims.html).

File: librosa/feature/rhythm.py [hunk 7]
```diff
@@ -139,6 +139,7 @@ def tempogram(
         raise ParameterError("win_length must be a positive integer")
 
     ac_window = get_window(window, win_length, fftbins=True)
+    ac_window = ac_window.reshape((-1,1)) # explicit broadcast
```
Comment (bmcfee) [inline librosa/feature/rhythm.py:143]: This needs to expand out to the full shape of the framed onset envelope, not just one leading dimension.

File: tests/test_multichannel.py [hunk 8]
```diff
@@ -59,6 +92,119 @@ def test_stft_multi(y_multi):
...
+
+    # Check each channel
+    assert np.allclose(D[0], D0)
+    assert np.allclose(D[1], D1)
+
+    # Check that they're not both the same
+    assert not np.allclose(D0, D1)
+
+
+def test_tempogram(y_multi):
```
Comment (bmcfee) [inline tests/test_multichannel.py:116]: this test should just use `s_multi` instead of y_multi`.

If `onset_envelope` is provided, then `y` will not be used in the tempogram calculations.

File: tests/test_multichannel.py [hunk 9]
```diff
@@ -59,6 +92,119 @@ def test_stft_multi(y_multi):
...
+
+    # Check each channel
+    assert np.allclose(t[0], t0)
+    assert np.allclose(t[1], t1)
+
+    # Check that they're not both the same
+    assert not np.allclose(t0, t1)
+
+
+def test_fourier_tempogram(y_multi):
```
Comment (bmcfee) [inline tests/test_multichannel.py:139]: same comment as above, we should use the `s_multi` fixture here

**Comment by bmcfee** [comment]
Thanks @iranroman !  This all looks good now, modulo a couple of small comments on the tests.

Only other remaining question is if this PR will include plp or not?  I'm fine with either way, just curious.

**Comment by iranroman** [comment]
Thanks for the feedback and guidance @bmcfee. 

> Only other remaining question is if this PR will include plp or not? 

I'll try to get plp done today (as well as the test changes)

File: tests/test_multichannel.py [hunk 10]
```diff
@@ -201,7 +203,64 @@ def test_tempo_multi(y_multi):
 
     # Check that they're not both the same
     assert not np.allclose(t0, t1)
-    print(t,t0,t1)
+
+
```
Comment (bmcfee) [inline tests/test_multichannel.py:207]: PLP Implementation looks good!

Some notes on the test:

1. It looks like `ctx` is always `dnr` (do not raise) here, in which case we should be able to skip it entirely.
2. One hop and window is sufficient here, no use in burning cycles for things we already test elsewhere
3. `use_onset` parameter is not used in the test.  I don't think it's critical that we test multi-channel with and without onset envelope construction since the code path is identical and we test that elsewhere.  May as well remove that parameter.

Comment (iranroman) [reply tests/test_multichannel.py:207]: Thanks!
    
    1. I was trying other parameter combinations but I removed some, and now there's only `dnr()`
    2. alright
    3.  same story as 1

Pushing a commit with fixes now.

**Review by bmcfee — APPROVED** [review]
Thanks!  This all looks great, happy to merge once CI finishes. Unless there's anything else to add in this PR?

**Comment by iranroman** [comment]
I think for `beat.py` this is it. You said on #1351 that we are only implementing `tempo` and `plp`. So those are done.
